console/main.cpp

```diff
@@ -122,9 +125,31 @@ struct App {
       });
     };
   }
-  ~App() { the_app = nullptr; }
+  ~App() { self() = nullptr; }
+
+  bool interrupt() { return interrupted.exchange(true); }
+
+  static void disable_interrupt_handle() { signal(SIGINT, SIG_DFL); }
+
+  static void setup_interrupt_handler() {
+    signal(
+        SIGINT, +[](int) {
+          self()->interrupt();
```

Comment:
I don't think we can do this, unfortunately -- there's [almost nothing](https://wiki.sei.cmu.edu/confluence/display/c/SIG30-C.+Call+only+asynchronous-safe+functions+within+signal+handlers#:~:text=For%20strictly%20conforming%20programs%2C%20only,from%20within%20a%20signal%20handler.&text=Implementations%20may%20define%20a%20list,called%20within%20a%20signal%20handler.) you can do in a `signal()` handler, which is one of the reasons why I didn't touch it. Calling a general function like `interrupt()` would be a recipe to invite folks to do one of the things you can't do...